src/Accelerators/NNPA/Transform/ZLow/ZLowFinalPassRewrite.cpp

```diff
@@ -0,0 +1,183 @@
+          directionAttr = userLstmOp.direction();
+          break;
+        }
+      }
+      if (!strcmp(opName, "zlow.gru")) {
+        directionAttr = llvm::dyn_cast<ZLowGRUOp>(user).direction();
+        break;
+      }
+    }
+    StringAttr prevLayerAttr;
+    if (directionAttr.empty() || !strcmp(directionAttr.data(), "")) {
+      prevLayerAttr = rewriter.getStringAttr("none");
+    } else if (!strcmp(directionAttr.data(), "bidirectional")) {
+      prevLayerAttr = rewriter.getStringAttr("bidir");
+    } else {
+      prevLayerAttr = rewriter.getStringAttr("uni");
+    }
+    // Emit a ZLow operation.
+    rewriter.create<ZLowLSTMOp>(lstmOp.getLoc(), lstmOp.input(), lstmOp.h0(),
+        lstmOp.c0(), lstmOp.input_weights(), lstmOp.input_bias(),
+        lstmOp.hidden_weights(), lstmOp.hidden_bias(), lstmOp.work_area(),
+        lstmOp.shape(), lstmOp.hn_output(), lstmOp.cf_output(),
+        lstmOp.directionAttr(), lstmOp.return_all_stepsAttr(), prevLayerAttr);
+    rewriter.eraseOp(lstmOp);
+
+    return success();
+  }
+};
+
+class SetPrevLayerInGRUOpPattern : public OpRewritePattern<ZLowGRUOp> {
+public:
+  using OpRewritePattern<ZLowGRUOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ZLowGRUOp gruOp, PatternRewriter &rewriter) const override {
+    Value gruInput = gruOp.input();
+    StringRef prevLayer = gruOp.prev_layer();
+    if (strcmp(prevLayer.data(), "not_set")) // if prev_layer is set already
+      return failure();
+
+    // Search for LSTM/GRU op that generates the input argument.
+    StringRef directionAttr = "";
+    for (Operation *user : gruInput.getUsers()) {
+      const char *opName = user->getName().getStringRef().data();
+      if (!strcmp(opName, "zlow.gru")) {
+        ZLowGRUOp userGruOp = llvm::dyn_cast<ZLowGRUOp>(user);
+        if (userGruOp != gruOp) {
+          directionAttr = userGruOp.direction();
+          break;
+        }
+      }
+      if (!strcmp(opName, "zlow.lstm")) {
+        directionAttr = llvm::dyn_cast<ZLowLSTMOp>(user).direction();
+        break;
+      }
+    }
+    StringAttr prevLayerAttr;
+    if (directionAttr.empty() || !strcmp(directionAttr.data(), "")) {
+      prevLayerAttr = rewriter.getStringAttr("none");
+    } else if (!strcmp(directionAttr.data(), "bidirectional")) {
+      prevLayerAttr = rewriter.getStringAttr("bidir");
+    } else {
+      prevLayerAttr = rewriter.getStringAttr("uni");
+    }
+    // Emit a ZLow operation.
+    rewriter.create<ZLowGRUOp>(gruOp.getLoc(), gruOp.input(), gruOp.h0(),
+        gruOp.input_weights(), gruOp.input_bias(), gruOp.hidden_weights(),
+        gruOp.hidden_bias(), gruOp.work_area(), gruOp.shape(),
+        gruOp.hn_output(), gruOp.directionAttr(), gruOp.return_all_stepsAttr(),
+        prevLayerAttr);
+    rewriter.eraseOp(gruOp);
+
+    return success();
+  }
+};
+
+/*!
+ *  Function pass that optimizes ZLowIR.
+ *  This Pass should be the final pass toupdate ZLowIR.
+ */
+class ZLowRewriteFinalPass
```

Comment:
Thank you very much for the comment. I agree that this is an important discussion point of this PR.

> I am not clear on the reason this pass cannot be run before other passes. If it is run before other passes it might not find the pattern it expects. 
We need to check if the input comes from the previous LSTM/GRU operation directly. I am afraid a situation that another pass updating zlow dialect removes operations generating the input of this operation after applying this pass. 

> Also, more importantly, can the 2 patterns at lines 170-171 be added to an existing pass rather than having to add them here ?
To avoid the described situation, the SetPrevLayerInLSTMOpPattern and SetPrevLayerInGRUOpPattern should be done after all possible passes those may remove operations generating the input of this operation.
An independent pass is necessary to guarantee the order of processing, according to my understanding.

Reply:
@etiotto Thank you very much for the suggestion. 
I deleted the ZLowRewriteFinalPass, and merged SetPrevLayerInLSTMOpPattern and SetPrevLayerInGRUOpPattern into ZLowRewritePass with lower benefit than the original patterns to schedule them after the original ones.